### PR TITLE
Edit Posts

### PR DESCRIPTION
**Pull request by nicholasvperry**
# Description
Added the ability to edit posts. When the user navigates to posts menu they have the ability to click the edit button. There is also an edit button inside the details view. User can edit post and save and view edited post.
Added MyPostDetails.cshtml and MyPosts.cshtml

#Type of changes
[Check Trello Ticket #8](https://trello.com/b/ECtaPfZr/the-daily-debuggers)
- [x] New feature (non-breaking change which adds functionality)

# Testing Instructions

1. git add and git commit all changes in your current branch.
2. git checkout main
3. git fetch -a
4. git checkout edit-post-np
5. Navigate to the root directory for this repo (directory-to)/tabloidmvc-the-daily-debuggers
6. Start TabloidMVC.sln
7. Make sure your db is created with seed data (Skip to 9 if this is already done)
8. Run [01_TabloidMVC_Create_DB.sql](https://github.com/NewForce-Cohort-5/tabloidmvc-the-daily-debuggers/blob/main/SQL/01_TabloidMVC_Create_DB.sql) and [02_TabloidMVC_Seed_Data.sql](https://github.com/NewForce-Cohort-5/tabloidmvc-the-daily-debuggers/blob/main/SQL/02_TabloidMVC_Seed_Data.sql) queries in Visual Studio
9. Start the server in Debug Mode F5
10. Log in as admin admin@example.com (if not already logged in as admin)
11. Navigate to the posts link
12. Select edit symbol beside post or click view post then edit symbol
13. Edit post
14. Click save
15. You will see the updated post

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have added test instructions that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by cbgriffith — APPROVED** [review]
Edit works perfectly.

File: TabloidMVC/Controllers/PostController.cs [hunk 1]
```diff
@@ -20,9 +24,22 @@ public PostController(IPostRepository postRepository, ICategoryRepository catego
             _categoryRepository = categoryRepository;
         }
 
+        //Posts sorted by PublishDateTime newest first
         public IActionResult Index()
         {
+            //Return posts
             var posts = _postRepository.GetAllPublishedPosts();
+            //Sort by PublishDateTime
+            posts.Sort((y, x) => DateTime.Compare((DateTime)x.PublishDateTime, (DateTime)y.PublishDateTime));
```
Comment (andy-birt) [inline TabloidMVC/Controllers/PostController.cs:33]: I don't think you need this line now that you're sorting the posts in the sql command.

File: TabloidMVC/Controllers/PostController.cs [hunk 2]
```diff
@@ -73,5 +92,82 @@ private int GetCurrentUserProfileId()
...
+
+                return RedirectToAction("Index");
+            }
+            catch (Exception ex)
+            {
+                return View(post);
+            }
+        }
+
+        // GET: PostController/Delete
```
Comment (andy-birt) [inline TabloidMVC/Controllers/PostController.cs:148]: Delete is a separate ticket. You may comment out the delete functionality and put it in a different pr. I'm sure it's already working.

File: TabloidMVC/Repositories/PostRepository.cs [hunk 3]
```diff
@@ -197,5 +200,61 @@ private Post NewPostFromReader(SqlDataReader reader)
...
+                    cmd.Parameters.AddWithValue("@categoryId", post.CategoryId);
+                    cmd.Parameters.AddWithValue("@userProfileId", post.UserProfileId);
+                    cmd.Parameters.AddWithValue("@id", post.Id);
+
+                    cmd.ExecuteNonQuery();
+                }
+            }
+        }
+
+        public void DeletePost(int postId)
```
Comment (andy-birt) [inline TabloidMVC/Repositories/PostRepository.cs:239]: same thing here as previous comment.

**Review by andy-birt — CHANGES_REQUESTED** [review]
It looks good. The only concern is mixing the delete ticket with the edit. We got to have a PR for each ticket.

**Review by andy-birt — APPROVED** [review]

**Review by jordan-castelloe — CHANGES_REQUESTED** [review]
Hey! Generally this looks great. I did find one bug-- the "Header Image URL" form field is not required, but if you leave that field blank I get the following exception: 
```
An unhandled exception occurred while processing the request.
InvalidOperationException: The model item passed into the ViewDataDictionary is of type 'TabloidMVC.Models.Post', but this ViewDataDictionary instance requires a model item of type 'TabloidMVC.Models.ViewModels.PostCreateViewModel'.
```
Your options are to either make that field required (if you actually want it to be required-- maybe discuss with the team!) or write in some conditional logic in your controller so that it doesn't try to insert null values into the database. When you've made the fix, let me know and I'll make a final pass!

**Comment by jordan-castelloe** [comment]
@nicholasvperry checking in on this-- did you get a chance to push up these changes? Want me to make another pass?

**Comment by nicholasvperry** [comment]
Fixed "Header Image URL" null issue. Edit will now allow with a null value.

**Review by jordan-castelloe — APPROVED** [review]

**Pull request merged**